lib/Target/OpenQASM/TranslateToOpenQASM.cpp

```diff
@@ -116,72 +111,94 @@ static LogicalResult emitOperation(Emitter &emitter, ModuleOp moduleOp) {
-  Value wireOrArray = allocaOp.getWireOrArray();
+static LogicalResult emitOperation(Emitter &emitter, quake::AllocaOp allocaOp) {
+  Value qrefOrQvec = allocaOp.getQrefOrVec();
   auto name = emitter.createName();
   auto size = 1;
-  if (auto array =
-          wireOrArray.getType().dyn_cast_or_null<qtx::WireArrayType>()) {
-    size = array.getSize();
+  if (auto qvec = dyn_cast<quake::QVecType>(qrefOrQvec.getType())) {
+    if (!qvec.hasSpecifiedSize())
+      return allocaOp.emitError("allocates unbounded qvec");
+    size = qvec.getSize();
   }
   emitter.os << llvm::formatv("qreg {0}[{1}];\n", name, size);
-  if (wireOrArray.getType().isa<qtx::WireType>())
+  if (isa<quake::QRefType>(qrefOrQvec.getType()))
     name.append("[0]");
-  emitter.getOrAssignName(wireOrArray, name);
+  emitter.getOrAssignName(qrefOrQvec, name);
   return success();
 }
 
-static LogicalResult emitOperation(Emitter &emitter, qtx::ApplyOp op) {
-  if (op.getNumClassicalResults())
+static LogicalResult emitOperation(Emitter &emitter, quake::ApplyOp op) {
+  // In quake, kernels only return classical types. Thus, we check whether the
+  // numbers of results is zero or not.
+  if (op.getNumResults() > 0)
     return op.emitError("cannot return classical results");
+  if (!op.getControls().empty())
+    return op.emitError("cannot add controls to a gate call");
   emitter.os << op.getCallee();
-  if (op.getNumParameters()) {
+
+  // Separate classical and quantum arguments.
+  SmallVector<Value, 4> parameters;
+  SmallVector<Value, 4> targets;
+  for (auto arg : op.getArgs()) {
+    if (isa<quake::QRefType, quake::QVecType>(arg.getType()))
+      targets.push_back(arg);
+    else
+      parameters.push_back(arg);
+  }
+  if (!parameters.empty()) {
     emitter.os << '(';
-    llvm::interleaveComma(op.getParameters(), emitter.os, [&](auto param) {
+    llvm::interleaveComma(parameters, emitter.os, [&](auto param) {
       emitter.os << emitter.getOrAssignName(param);
     });
     emitter.os << ')';
   }
   emitter.os << ' ';
-  llvm::interleaveComma(op.getTargets(), emitter.os, [&](auto target) {
+  llvm::interleaveComma(targets, emitter.os, [&](auto target) {
     emitter.os << emitter.getOrAssignName(target);
   });
-  emitter.mapValuesName(op.getTargets(), op.getNewTargets());
   emitter.os << ";\n";
   return success();
 }
 
-static LogicalResult emitOperation(Emitter &emitter, qtx::CircuitOp circuitOp) {
-  // Check if we can translate
-  if (circuitOp.getClassicalResultTypes().size())
-    return circuitOp.emitError("cannot return classical results");
-  for (auto target : circuitOp.getTargets()) {
-    if (isa<qtx::WireArrayType>(target.getType()))
-      return circuitOp.emitError(
-          "Cannot translate array arguments into OpenQASM 2.0");
+static LogicalResult emitOperation(Emitter &emitter, func::FuncOp op) {
+  // In quake, kernels only return classical types. Thus, we check whether the
```

Comment:
It'll be permitted to transform a device kernel into "pass-by-value" form with wire arguments and wire results. There isn't a pass that does that, but we can certainly write it. In fact, I'd like to see if the !quake.control type can't be leveraged as well with such a transformation to help detect circuit bugs.